### PR TITLE
TW-1032: Language settings

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket:
- #1032

#### Mobile 

https://github.com/linagora/twake-on-matrix/assets/99852347/23f18d2f-9023-4406-99d6-69d6eca0284d

#### Web

https://github.com/linagora/twake-on-matrix/assets/99852347/ef3cd29c-f1fb-4099-83a6-3a799325e44a

### CODE REVIEW

File: lib/di/global/get_it_initializer.dart [hunk 1]
```diff
@@ -85,6 +91,7 @@ class GetItInitializer {
     NetworkConnectivityDI().bind();
     getIt.registerSingleton(ResponsiveUtils());
     getIt.registerSingleton(TwakeEventDispatcher());
+    getIt.registerSingleton(LanguageCacheManager());
```
Comment (hoangdat) [inline lib/di/global/get_it_initializer.dart:94]: Why we need it as a singleton?

File: lib/di/global/get_it_initializer.dart [hunk 2]
```diff
@@ -129,6 +141,11 @@ class GetItInitializer {
         getIt.get<MediaAPI>(),
       ),
     );
+    getIt.registerLazySingleton(
```
Comment (hoangdat) [inline lib/di/global/get_it_initializer.dart:144]: Idem

File: lib/di/global/get_it_initializer.dart [hunk 3]
```diff
@@ -209,5 +231,10 @@ class GetItInitializer {
       () => ChatGetPinnedEventsInteractor(),
     );
     getIt.registerSingleton<ContactsManager>(ContactsManager());
+    getIt.registerSingleton<SaveLanguageInteractor>(
```
Comment (hoangdat) [inline lib/di/global/get_it_initializer.dart:234]: Idem

**Comment by sherlockvn** [comment]
Does language really need to cache ? Because as the user, They usually use one, or at most 2 languages, and I think they don't switch language often

**Comment by hoangdat** [comment]
And one question, we can synchronize the language to other clients or not? Fx: you have browser, you change language in browser, and your phone app update too.?

**Comment by nqhhdev** [comment]
> And one question, we can synchronize the language to other clients or not? Fx: you have browser, you change language in browser, and your phone app update too.?

1. If the priority exists locally, it will be used, otherwise it will be used by the system.
2. If you saved to local when changing the language on the system, it will not be applied because we are prioritizing local use

File: assets/l10n/intl_en.arb [hunk 4]
```diff
@@ -2927,5 +2927,30 @@
...
   "fetchingPhonebookContacts": "Fetching contacts from device...({progress}% completed)",
-  "activeNow": "Active now"
+  "activeNow": "Active now",
+  "languageEnglish": "English",
+  "@languageEnglish": {
+    "type": "text",
+    "placeholders_order": [],
+    "placeholders": {}
+  },
+  "languageVietnamese": "Vietnamese",
```
Comment (drminh2807) [inline assets/l10n/intl_en.arb:2937]: you should use this lib to get local name from local codes
https://pub.dev/packages/flutter_localized_locales
we also need support all languages defined in `L10n.supportedLocales`

Comment (nqhhdev) [reply assets/l10n/intl_en.arb:2937]: Don't need to use lib. Something so simple depends on an overly complicated lib

Comment (hoangdat) [reply assets/l10n/intl_en.arb:2937]: I prefer to use lib in this case. It is simple now, but it can more complicate in the future.

**Review by drminh2807 — APPROVED** [review]

File: lib/data/local/localizations/language_cache_manager.dart [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import 'dart:ui';
+import 'package:fluffychat/utils/famedlysdk_store.dart';
+import 'package:matrix/matrix.dart';
+
+class LanguageCacheManager {
+  static const keyLanguageCode = 'LANGUAGE_CODE';
+  static const keyCountryCode = 'COUNTRY_CODE';
+
+  final Store pres = Store();
```
Comment (sherlockvn) [inline lib/data/local/localizations/language_cache_manager.dart:9]: Can you clarify what is this, i think you should change name. and should it be singleton ?

Comment (nqhhdev) [reply lib/data/local/localizations/language_cache_manager.dart:9]: It belongs to matrix sdk, I reuse it all the time

Comment (hoangdat) [reply lib/data/local/localizations/language_cache_manager.dart:9]: please use getIt.

Comment (nqhhdev) [reply lib/data/local/localizations/language_cache_manager.dart]: done

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**